File: Algorithms/p047_Permutations_II/p047_Permutations_II_2.py
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        if not nums:
            return []
        nums.sort()
        return self.permuteUniqueHelper(nums)

    def permuteUniqueHelper(self, nums):
        if not nums:
            return [[]]
        res = []
        for i in range(len(nums)):
            if i>0 and nums[i]==nums[i-1]:
                continue
            nums1 = nums[0:i]+nums[i+1:]
            res1 = self.permuteUniqueHelper(nums1)
            for e in res1:
                e.insert(0,nums[i])
            res += res1
        return res

File: Algorithms/p047_Permutations_II/p047_Permutations_II_2.cpp
class Solution {
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        return permuteUniqueHelper(nums);
    }
    vector<vector<int>> permuteUniqueHelper(vector<int>& nums) {
        if(nums.empty()) return {{}};
        vector<vector<int>> res;
        for(int i=0;i<nums.size();++i){
            if(i>0&&nums[i]==nums[i-1]) continue;
            vector<int> nums1(nums.begin(),nums.begin()+i);
            nums1.insert(nums1.end(),nums.begin()+i+1,nums.end());
            vector<vector<int>> res1 = permuteUniqueHelper(nums1);
            for(auto &e : res1){
                e.insert(e.begin(),nums[i]);
            }
            res.insert(res.end(),make_move_iterator(res1.begin()),make_move_iterator(res1.end()));
        }
        return res;
    }
};
